### PR TITLE
Map update in accordance to ROS time.

### PR DESCRIPTION
**Pull request by AzaelCicero**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #3325 |
| Primary OS tested on | (Ubuntu) |
| Robotic platform tested on | autonomy driving real data replayed with rosbag |

---

## Description of contribution in a few bullet points

* Changed update map frequency control mechanism from WallRate to ROS Timer, so now it respects simulation time if use_sim_time flag is used.

## Description of documentation updates required from your changes

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

**Comment by SteveMacenski** [comment]
@AlexeyMerzlyakov please review in contrast to your suggestion

File: nav2_costmap_2d/src/costmap_2d_ros.cpp [hunk 1]
```diff
@@ -447,68 +449,52 @@ Costmap2DROS::getOrientedFootprint(std::vector<geometry_msgs::msg::Point> & orie
...
-        if ((last_publish_ + publish_cycle_ < current_time) ||  // publish_cycle_ is due
-          (current_time <
-          last_publish_))      // time has moved backwards, probably due to a switch to sim_time // NOLINT
-        {
-          RCLCPP_DEBUG(get_logger(), "Publish costmap at %s", name_.c_str());
-          costmap_publisher_->publishCostmap();
+    auto current_time = now();
 
-          for (auto & layer_pub: layer_publishers_) {
-            layer_pub->publishCostmap();
```
Comment (SteveMacenski) [inline nav2_costmap_2d/src/costmap_2d_ros.cpp:492]: Ditto on layer publishers

File: nav2_costmap_2d/src/costmap_2d_ros.cpp [hunk 2]
```diff
@@ -309,8 +312,7 @@ Costmap2DROS::on_deactivate(const rclcpp_lifecycle::State & /*state*/)
   stop();
 
   // Map thread stuff
-  map_update_thread_shutdown_ = true;
-  map_update_thread_->join();
+  map_update_timer_ = nullptr;
```
[the remark this answers is not in the document]
Comment (AzaelCicero) [reply nav2_costmap_2d/src/costmap_2d_ros.cpp]: Ok, I am going to use cancel/reset in appropriate places.

File: nav2_costmap_2d/src/costmap_2d_ros.cpp [hunk 3]
```diff
@@ -447,68 +452,61 @@ Costmap2DROS::getOrientedFootprint(std::vector<geometry_msgs::msg::Point> & orie
 }
 
 void
-Costmap2DROS::mapUpdateLoop(double frequency)
+Costmap2DROS::mapUpdateLoop()
 {
-  RCLCPP_DEBUG(get_logger(), "mapUpdateLoop frequency: %lf", frequency);
+  static const double expected_period = 1. / map_update_frequency_;
```
Comment (AlexeyMerzlyakov) [inline nav2_costmap_2d/src/costmap_2d_ros.cpp:457]: `map_update_frequency_` should be checked to be `!= 0.0` to avoid division by zero

File: nav2_costmap_2d/src/costmap_2d_ros.cpp [hunk 4]
```diff
@@ -279,9 +279,15 @@ Costmap2DROS::on_activate(const rclcpp_lifecycle::State & /*state*/)
...
-  map_update_thread_ = std::make_unique<std::thread>(
-    std::bind(&Costmap2DROS::mapUpdateLoop, this, map_update_frequency_));
+
+  if (map_update_frequency_ > 0.0 && map_update_timer_ == nullprt) {
+    map_update_timer_ = rclcpp::create_timer(
+      this, get_clock(), rclcpp::Duration::from_seconds(1 / map_update_frequency_),
+      [this]() {mapUpdateLoop();});
+  }
+  else {
+    map_update_timer_->reset();
```
[the remark this answers is not in the document]
Comment (AzaelCicero) [reply nav2_costmap_2d/src/costmap_2d_ros.cpp:289]: Right, I have rewritten the conditions to prevent it.

File: nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_ros.hpp [hunk 5]
```diff
@@ -343,12 +343,10 @@ class Costmap2DROS : public nav2_util::LifecycleNode
   /**
    * @brief Function on timer for costmap update
    */
-  void mapUpdateLoop(double frequency);
-  bool map_update_thread_shutdown_{false};
+  void mapUpdateLoop();
   std::atomic<bool> stop_updates_{false};
-  std::atomic<bool> initialized_{false};
   std::atomic<bool> stopped_{true};
```
[the remark this answers is not in the document]
Comment (AzaelCicero) [reply nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_ros.hpp:348]: No, we do not need this one.

File: nav2_costmap_2d/src/costmap_2d_ros.cpp (hunk 3 above)
Comment (AzaelCicero) [reply nav2_costmap_2d/src/costmap_2d_ros.cpp:457]: ok

File: nav2_costmap_2d/src/costmap_2d_ros.cpp [hunk 6]
```diff
@@ -447,68 +452,61 @@ Costmap2DROS::getOrientedFootprint(std::vector<geometry_msgs::msg::Point> & orie
 }
 
 void
-Costmap2DROS::mapUpdateLoop(double frequency)
+Costmap2DROS::mapUpdateLoop()
 {
-  RCLCPP_DEBUG(get_logger(), "mapUpdateLoop frequency: %lf", frequency);
```
[the remark this answers is not in the document]
Comment (AzaelCicero) [reply nav2_costmap_2d/src/costmap_2d_ros.cpp:452]: This log entry at this place does not make any longer sense. This code will be executed each time the timer fall. I moved it around the creation of the timer.

File: nav2_costmap_2d/src/costmap_2d_ros.cpp [hunk 7]
```diff
@@ -279,9 +280,22 @@ Costmap2DROS::on_activate(const rclcpp_lifecycle::State & /*state*/)
   // Create a thread to handle updating the map
   stopped_ = true;  // to active plugins
   stop_updates_ = false;
-  map_update_thread_shutdown_ = false;
-  map_update_thread_ = std::make_unique<std::thread>(
-    std::bind(&Costmap2DROS::mapUpdateLoop, this, map_update_frequency_));
+
+  if (map_update_timer_ == nullptr) {
```
Comment (SteveMacenski) [inline nav2_costmap_2d/src/costmap_2d_ros.cpp:284]:
```suggestion
  if (!map_update_timer_) {
```

File: nav2_costmap_2d/src/costmap_2d_ros.cpp [hunk 8]
```diff
@@ -309,8 +323,9 @@ Costmap2DROS::on_deactivate(const rclcpp_lifecycle::State & /*state*/)
   stop();
 
   // Map thread stuff
-  map_update_thread_shutdown_ = true;
-  map_update_thread_->join();
+  if (map_update_timer_ != nullptr) {
```
Comment (SteveMacenski) [inline nav2_costmap_2d/src/costmap_2d_ros.cpp:326]:
```suggestion
  if (map_update_timer_) {
```

File: nav2_costmap_2d/src/costmap_2d_ros.cpp [hunk 9]
```diff
@@ -562,13 +547,6 @@ Costmap2DROS::start()
     stopped_ = false;
   }
   stop_updates_ = false;
-
-  // block until the costmap is re-initialized.. meaning one update cycle has run
-  rclcpp::Rate r(20.0);
-  while (rclcpp::ok() && !initialized_) {
```
[the remark this answers is not in the document]
Comment (SteveMacenski) [reply nav2_costmap_2d/src/costmap_2d_ros.cpp:568]: I think we have the `initialized_` member to ensure we trigger at least 1 time before we do other things which has been removed. This ensures that we have some data to look at and not blank-defaults if a request comes in sooner after startup and/or there is a very slow update rate. So to remove that, we need to make sure updateMap is triggered at some point during the lifecycle activation state (or readd).

File: nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_ros.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (AlexeyMerzlyakov) [reply nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d_ros.hpp:348]: I am not talking about `initialized_` removal. In a ROS2 default execution - is a single thread executor (https://github.com/ros2/rclcpp/blob/rolling/rclcpp/src/rclcpp/executors.cpp#L33), so two or more callback are guaranteed to be executed sequentially. Thus, if `std::thread` usage was refused, we do not need `std::atomic`  for remaining boolean indicators anymore:
``` 
  std::atomic<bool> stop_updates_{false};
  std::atomic<bool> initialized_{false};
  std::atomic<bool> stopped_{true};
```
-> might be
```
bool stop_updates_{false};
bool stopped_{true};
```

File: nav2_costmap_2d/src/costmap_2d_ros.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (AlexeyMerzlyakov) [reply nav2_costmap_2d/src/costmap_2d_ros.cpp:568]: As I see from the code, currently `initialized_` indicator is being used to prevent uninitialized map usage from `Costmap2DROS::start()` and `Costmap2DROS::resume()`.  The `start()` routine is being called from `Costmap2DROS::on_activate()` callback (or `start()` also could be called manually, as in public domain).
In both cases in `start()` and in `resume()`, we are waiting in for a `mapUpdateLoop()` loop will call `updateMap()` and thus  set `initialized_` to `true`. Instead of this we could remove these waits and just manually call `mapUpdateLoop()` routine, which will set all maps.

There are remain setting of `initialized_ = false` in `stop()` and `pause()` routines. If `pause()` was called, `initialized_` variable is used only in `start()` and in `resume()`, where we will intentionally update maps instead, so this is also not an issue.
As for the `stop()` is being called, `initalized_` is set to `false` and we need re-initialization. But since is `stopped_ = false` too, `mapUpdateLoop()` will be skipped and maps are not updated. This is not an issue, if call `start()`, but won't work for `resume()` routine.

In other words, possible pitfall we could get is: call `resume()` after `stop()` if apply this patch. @SteveMacenski, did you mean this case, or I've missed something else?

**Comment by AlexeyMerzlyakov** [comment]
@AzaelCicero, how the things are going? Any updates after reviews on this ticket?

**Comment by SteveMacenski** [comment]
https://github.com/ros-planning/navigation2/pull/3404 supersedes
